Sprint 14

nurboltt opened the pull request.

nurboltt committed:
1s iteration done

nurboltt committed:
delete ViewController

nurboltt committed:
моковая коллекция отображается

nurboltt committed:
fixed all bugs

So1ovyov commented on Tracker/AppDelegate.swift:
```diff
@@ -10,10 +10,13 @@ import UIKit
-
+    var window: UIWindow?
+    
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
```
Предлагаю длинные конструкции, такие как init-ы/методы и другие, располагать сверху-вниз параметр под параметром. Это облегчает чтение и восприятие кода, в отличии от одной длинной строки:
```
    func application(
        _ application: UIApplication,
        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    ) -> Bool {
        window = UIWindow()
        window?.rootViewController = TabBarController()
        window?.makeKeyAndVisible()
        return true
    }

    // MARK: UISceneSession Lifecycle

    func application(
        _ application: UIApplication,
        configurationForConnecting connectingSceneSession: UISceneSession,
        options: UIScene.ConnectionOptions
    ) -> UISceneConfiguration {
        UISceneConfiguration(name: "Default Configuration", sessionRole: connectingSceneSession.role)
    }

    func application(
        _ application: UIApplication,
        didDiscardSceneSessions sceneSessions: Set<UISceneSession>
    ) {}
```

So1ovyov commented on Tracker/AppDelegate.swift:
```diff
@@ -10,10 +10,13 @@ import UIKit
+    var window: UIWindow?
+    
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
```
Также комментарии и дефолтную докеументацию в методах можно также удалить.

So1ovyov commented on Tracker/CategoryTableViewCell.swift:
```diff
@@ -0,0 +1,52 @@
+final class CategoryTableViewCell: UITableViewCell {
+    
+    static let categoryCellIdentifier = "categoryCellIdentifier"
+    private lazy var titleLabel: UILabel = {
```
Точно ли есть необходимость в ленивой переменной ?
Давай используем константу:
```
private let titleLabel
```

So1ovyov commented on Tracker/CategoryTableViewCell.swift:
```diff
@@ -0,0 +1,52 @@
+        backgroundColor = UIColor(named: "yp-light-gray")
+        contentView.layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
+        
+        [titleLabel].forEach {
```
`forEach` обычно используют когда несколько элементов, а тут у тебя массив из одного элемента.
Не критично, но нужно ли ?

So1ovyov commented on Tracker/CategoryTableViewCell.swift:
```diff
@@ -0,0 +1,52 @@
+        NSLayoutConstraint.activate([
+            titleLabel.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 16),
+            titleLabel.centerYAnchor.constraint(equalTo: contentView.centerYAnchor),
+            
```
Лишняя пустая строка, давай почистим.

So1ovyov commented on Tracker/CategoryTableViewCell.swift:
```diff
@@ -0,0 +1,52 @@
+        ])
+    }
+    
+    func configure(with title: String) {
```
Вот это отличное решение, свойство инкапсулировано, а чтобы в нем что-то изменить создана ручка, молодец!

So1ovyov commented on Tracker/ColorCollectionViewCell.swift:
```diff
@@ -0,0 +1,12 @@
+import UIKit
+
+final class ColorCollectionViewCell: UICollectionViewCell {
+    // TODO: Add colorCollectionViewCell in 15 sprint
```
Хорошая практика оставлять `TODO`, молодец!

So1ovyov commented on Tracker/Controllers/AddTrackerViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    
+    var trackerSelectedClosure: ((Tracker) -> Void)?
+    
+    private lazy var habitButton: UIButton = {
```
Давай все подобные свойства сделаем константами, кажется что в ленивых вычислениях нет смысла, так как при инициализации `ViewController`-а ты сразу используешь эти UI элементы в нем, тут нет никакой оптимизации `lazy`

So1ovyov commented on Tracker/Controllers/AddTrackerViewController.swift:
```diff
@@ -0,0 +1,100 @@
+    }
+    
+    @objc func irregularButtonTapped(_ sender: UIButton) {
+        let newHabitViewController = NewHabitViewController(titleText: "Новое нерегулярное событие", cells: [.category("Важное")])
```
Очень круто, что ты сделал настраиваемый `NewHabitViewController` прям отдельный респект, за переиспользуемую сущность.

So1ovyov commented on Tracker/Controllers/CategoryTableViewController.swift:
```diff
@@ -0,0 +1,120 @@
+
+import UIKit
+
+enum FakeCategories: String, CaseIterable {
```
Давай придерживаться принципа 1 файл = 1 структура/класс/модель/enum. Не стоит раздувать файл и располагать в нем все сразу.

So1ovyov commented on Tracker/Controllers/NewHabitViewController.swift:
```diff
@@ -0,0 +1,301 @@
+
+import UIKit
+
+enum TableViewCellType {
```
Аналогично и здесь и в других классах, давай вынесем сущности в отдельные файлы.

So1ovyov commented on Tracker/Controllers/NewHabitViewController.swift:
```diff
@@ -0,0 +1,301 @@
+    private let tableView = UITableView(frame: .zero, style: .insetGrouped)
+    private var cells: [TableViewCellType]
+    private var newSchedule: [WeekDay] = []
+    private let emoji: [String] = [
```
Предлагаю все захардкоженные значения вынести в `enum Constants`, а моковые данные в `enum MockData`, это поможет разгрузить твои классы и хранить такие данные в одном месте.
В данном случае можно вынести это в `enum Constants`
```
enum Constants {
   static let emoji: [String] = []
}
```

So1ovyov commented on Tracker/Controllers/NewHabitViewController.swift:
```diff
@@ -0,0 +1,301 @@
+            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: 0),
+            tableView.heightAnchor.constraint(equalToConstant: 200),
+            
+            // TODO: Добавить в 15 спринте
```
Закомментированный код можно удалить, если он не используется

So1ovyov commented on Tracker/Controllers/NewHabitViewController.swift:
```diff
@@ -0,0 +1,301 @@
+    }
+    
+    @objc private func cancelButtonTapped() {
+        print("cancel button tapped")
```
Print-ы для дебага тоже можно удалить.

So1ovyov commented on Tracker/Controllers/NewHabitViewController.swift:
```diff
@@ -0,0 +1,301 @@
+        }
+        
+        let isIrregular = cells.firstIndex {
+            if case .schedule = $0 { return true }
```
Для подобной логики можно применить и `guard`
```
            guard case .schedule = $0 else { return false }
            return true
```

So1ovyov commented on Tracker/Controllers/NewHabitViewController.swift:
```diff
@@ -0,0 +1,301 @@
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        if indexPath.row == 0 {
+            // TODO: Добавить в 15 спринте
```
Давай почистим проект от закомментированного кода

So1ovyov commented on Tracker/Controllers/ScheduleTableViewController.swift:
```diff
@@ -0,0 +1,165 @@
+
+import UIKit
+
+enum WeekDay: String, CaseIterable {
```
В отдельный файл лучше вынести. Можно все эти `enum`-ы перенести в папку `Models`

So1ovyov commented on Tracker/EmojiCollectionView.swift:
```diff
@@ -0,0 +1,34 @@
+
+final class EmojiCollectionView: UICollectionReusableView {
+    
+    let titleLabel = UILabel()
```
А почему здесь ушел от концепции инкапсуляции элементов ?
```
private let titleLabel = UILabel()
```
Давай добавим метод:
```

    func update(text: String) {
        titleLabel.text = text
    }
```
А там где меняется текст поменять на использование метода этого:
```
view.update(text: visibleCategories[indexPath.section].title)
```

nurboltt replied:
titleLabel используется в TrackersViewController поэтому он не private

So1ovyov commented on Tracker/EmojiCollectionViewCell.swift:
```diff
@@ -0,0 +1,28 @@
+final class EmojiCollectionViewCell: UICollectionViewCell {
+    
+    static let emojiCellIdentifier = "emojiCell"
+    let titleLabel = UILabel()
```
Аналогично и здесь, раз уже начал делать все в одном стиле, то давай и эти небольшие ячейки приведем в порядок:
```
    private let titleLabel = UILabel()

    func update(text: String) {
        titleLabel.text = text
    }
```

nurboltt replied:
тут аналогично

So1ovyov commented on Tracker/ScheduleTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+        switcher.addTarget(self, action: #selector(switchToggled), for: .valueChanged)
+       return switcher
+    }()
+    var onSwitchToggle: ((Bool) -> Void)?
```
Если вдруг захочется как-то упросить конструкцию и сделать ее проще для восприятия, то можно использовать `typealias`, например создать где-нибудь в папке `Models` отдельный файл под `typealias` и для подобных комплишенов написать несколько конструкций:
```
typealias VoidClosure = (() -> Void)
typealias BoolClosure = ((Bool) -> Void)
typealias StringClosure = ((String) -> Void)
typealias IntClosure = ((Int) -> Void)
```
И таким образом в 25-ой строке можно использовать следующую конструкцию:
```
var onSwitchToggle: BoolClosure?
```
Согласись, что так выглядит немного проще, поэтому если будет желание можешь воспользоваться.

So1ovyov commented on Tracker/ScheduleTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    @objc private func switchToggled(_ sender: UISwitch) {
+           onSwitchToggle?(sender.isOn)
+       }
```
Сместились отступы, нужно поправить.

So1ovyov commented on Tracker/TrackerCell.swift:
```diff
@@ -0,0 +1,174 @@
+        with tracker: Tracker,
+        isCompletedToday: Bool,
+        completedDays: Int,
+        indexPath: IndexPath) {
```
Общий стиль обычно такой:
```
    func configure(
        with tracker: Tracker,
        isCompletedToday: Bool,
        completedDays: Int,
        indexPath: IndexPath
    ) {
        self.trackerId = tracker.id
        self.isCompletedToday = isCompletedToday
        self.indexPath = indexPath
```

So1ovyov commented on Tracker/TrackerCell.swift:
```diff
@@ -0,0 +1,174 @@
+            let wordDay = pluralizeDays(completedDays)
+            counterDayLabel.text = "\(wordDay)"
+            
+            plusButton.alpha = isCompletedToday ? 0.3 : 1.0
```
Круто, что не забываешь про тернарный оператор

So1ovyov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,390 @@
+    private var categories: [TrackerCategory] = []
+    private var visibleCategories: [TrackerCategory] = []
+    private var completedTrackers: [TrackerRecord] = []
+    var currentDate: Date = Date()
```
Кажется забыл инкапсулировать:
```
private var currentDate: Date = Date()
```

So1ovyov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,390 @@
+    
+    private let countNumber: CGFloat = 2
+  
+    private let colors = [ UIColor(named: "coll-green"), UIColor(named: "coll-orange")]
```
А в данной строке нет необходимости, так как ты константу `colors` не используешь в классе.

So1ovyov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,390 @@
+    
+    @objc private func addTask() {
+        let addTrackerViewController = AddTrackerViewController(titleText: "Создание трекера")
+//        addTrackerViewController.trackerSelectedClosure = { [weak self] tracker in
```
Закомментированный код можно удалить, если он не используется

So1ovyov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,390 @@
+                }
+            }
+            
+            if trackers.isEmpty {
```
Предлагаю немного сократить вложенность в подобных конструкциях с помощью `guard`-а
```
guard !trackers.isEmpty else { return nil }
```

So1ovyov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,390 @@
+            let cell = collectionView.dequeueReusableCell(
+                withReuseIdentifier: TrackerCell.trackerCellIdentifier,
+                for: indexPath
+            ) as? TrackerCell
```
Тут можно немного по другому расположить, чтобы это выглядело как единая конструкция:
```
        guard let cell = collectionView.dequeueReusableCell(
            withReuseIdentifier: TrackerCell.trackerCellIdentifier,
            for: indexPath
        ) as? TrackerCell else { return UICollectionViewCell() }
```

So1ovyov commented:
@nurboltt у тебя получился отличный проект, по сути все мои комментарии в ПР-е касаются либо инкапсуляции, либо расположения длинных конструкций, либо по поводу принтов и закомментированного кода, предлагаю это все исправить и довести проект до идеального состояния.
Также есть несколько моментов касаемо ТЗ и дизайна, более подробную информацию оставил в ревьюелке.
Жду исправлений :)

nurboltt committed:
исправлены замечания
datePicker оставил так как наставники сказали что любой формат даты годится, главное compact

So1ovyov commented:
@nurboltt очень оперативно и круто поправил все замечания, но осталась одна критическая ошибка. Обрати внимание на иконку `Done`, более подробно в ревьюелке описал.

nurboltt committed:
fixed

So1ovyov commented:
@nurboltt отличная работа, все замечания исправлены, ты очень хорошо потрудился и оперативно все закрыл, молодец!
Успехов в следующем спринте :)
